### PR TITLE
[12.0] [MIG] project_department

### PR DESCRIPTION
**Pull request by Nikul-Chaudhary**
Project Department Categorization
=================================

Add Department to Projects and task to corresponding tree, search and form views.

### CODE REVIEW

**Comment by Nikul-Chaudhary** [comment]
@pedrobaeza I added  dependencies https://github.com/OCA/project/pull/465/files#diff-58bf4ef7bb17e0cabe9c102156155697R1
But Travis still failing, Can you please help here

**Comment by pedrobaeza** [comment]
oca_dependencies.txt is not correct. The repository URL should be the forked one (the one that contains the branch proposed in the PR), not OCA.

File: project_department/README.rst [hunk 1]
```diff
@@ -0,0 +1,61 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
```
Comment (tarteo) [inline project_department/README.rst:1]: Use the seperated readme

File: project_department/__manifest__.py [hunk 2]
```diff
@@ -0,0 +1,17 @@
+# Copyright 2014 Joël Grand-Guillaume (Camptocamp)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
```
Comment (tarteo) [inline project_department/__manifest__.py]: Change the scheme to https (and in other files)

**Review by tarteo — CHANGES_REQUESTED** [review]

**Comment by Nikul-Chaudhary** [comment]
@tarteo Done

**Comment by Nikul-Chaudhary** [comment]
@tarteo @pedrobaeza Could you please review it?

**Review by alexey-pelykh — CHANGES_REQUESTED** [review]
Build fails, what to review here?

**Comment by Nikul-Chaudhary** [comment]
> Build fails, what to review here?

Opps, It's Base on https://github.com/OCA/account-analytic/pull/209

So fields not get, But you can review only Code, If base modules are Done then testing in runbot

**Comment by xabispacebiker** [comment]
I get the following error when installing project_department in Odoo12 in a test environment.

```
Odoo Server Error
Traceback (most recent call last):
  File "/opt/odoo12/odoo/models.py", line 1108, in _validate_fields
    check(self)
  File "/opt/odoo12/odoo/addons/base/models/ir_ui_view.py", line 351, in _check_xml
    self.postprocess_and_fields(view.model, view_doc, view.id)
  File "/opt/odoo12/odoo/addons/base/models/ir_ui_view.py", line 1092, in postprocess_and_fields
    self.raise_view_error(message, view_id)
  File "/opt/odoo12/odoo/addons/base/models/ir_ui_view.py", line 545, in raise_view_error
    raise ValueError(message)
ValueError: El campo `project_department_id` no existe

Contexto del error:
Vista `project.task.department.form`
[view_id: 2138, xml_id: project_department.view_task_form2, model: project.task, parent_id: 403]

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/odoo12/odoo/tools/convert.py", line 757, in parse
    self._tags[rec.tag](rec, de, mode=mode)
  File "/opt/odoo12/odoo/tools/convert.py", line 662, in _tag_record
    record = model.with_context(rec_context)._load_records([data], self.mode == 'update')
  File "/opt/odoo12/odoo/models.py", line 3788, in _load_records
    data['record'].write(data['values'])
  File "/opt/odoo12/addons/website/models/ir_ui_view.py", line 41, in write
    return super(View, self).write(vals)
  File "/opt/odoo12/odoo/addons/base/models/ir_ui_view.py", line 441, in write
    return super(View, self).write(self._compute_defaults(vals))
  File "/opt/odoo12/odoo/models.py", line 3306, in write
    fields[0].determine_inverse(records)
  File "/opt/odoo12/odoo/fields.py", line 1097, in determine_inverse
    getattr(records, self.inverse)()
  File "/opt/odoo12/odoo/addons/base/models/ir_ui_view.py", line 275, in _inverse_arch
    view.write(data)
  File "/opt/odoo12/addons/website/models/ir_ui_view.py", line 41, in write
    return super(View, self).write(vals)
  File "/opt/odoo12/odoo/addons/base/models/ir_ui_view.py", line 441, in write
    return super(View, self).write(self._compute_defaults(vals))
  File "/opt/odoo12/odoo/models.py", line 3269, in write
    self._write(store_vals)
  File "/opt/odoo12/odoo/models.py", line 3413, in _write
    self._validate_fields(vals)
  File "/opt/odoo12/odoo/models.py", line 1112, in _validate_fields
    raise ValidationError("%s\n\n%s" % (_("Error while validating constraint"), tools.ustr(e)))
odoo.exceptions.ValidationError: ('Error mientras se validaban las restricciones\n\nEl campo `project_department_id` no existe\n\nContexto del error:\nVista `project.task.department.form`\n[view_id: 2138, xml_id: project_department.view_task_form2, model: project.task, parent_id: 403]', None)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/odoo12/odoo/http.py", line 654, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/opt/odoo12/odoo/http.py", line 312, in _handle_exception
    raise pycompat.reraise(type(exception), exception, sys.exc_info()[2])
  File "/opt/odoo12/odoo/tools/pycompat.py", line 87, in reraise
    raise value
  File "/opt/odoo12/odoo/http.py", line 696, in dispatch
    result = self._call_function(**self.params)
  File "/opt/odoo12/odoo/http.py", line 344, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/opt/odoo12/odoo/service/model.py", line 97, in wrapper
    return f(dbname, *args, **kwargs)
  File "/opt/odoo12/odoo/http.py", line 337, in checked_call
    result = self.endpoint(*a, **kw)
  File "/opt/odoo12/odoo/http.py", line 939, in __call__
    return self.method(*args, **kw)
  File "/opt/odoo12/odoo/http.py", line 517, in response_wrap
    response = f(*args, **kw)
  File "/opt/odoo12/addons/web/controllers/main.py", line 966, in call_button
    action = self._call_kw(model, method, args, {})
  File "/opt/odoo12/addons/web/controllers/main.py", line 954, in _call_kw
    return call_kw(request.env[model], method, args, kwargs)
  File "/opt/odoo12/odoo/api.py", line 749, in call_kw
    return _call_kw_multi(method, model, args, kwargs)
  File "/opt/odoo12/odoo/api.py", line 736, in _call_kw_multi
    result = method(recs, *args, **kwargs)
  File "<decorator-gen-67>", line 2, in button_immediate_upgrade
  File "/opt/odoo12/odoo/addons/base/models/ir_module.py", line 73, in check_and_log
    return method(self, *args, **kwargs)
  File "/opt/odoo12/odoo/addons/base/models/ir_module.py", line 613, in button_immediate_upgrade
    return self._button_immediate_function(type(self).button_upgrade)
  File "/opt/odoo12/odoo/addons/base/models/ir_module.py", line 552, in _button_immediate_function
    modules.registry.Registry.new(self._cr.dbname, update_module=True)
  File "/opt/odoo12/odoo/modules/registry.py", line 86, in new
    odoo.modules.load_modules(registry._db, force_demo, status, update_module)
  File "/opt/odoo12/odoo/modules/loading.py", line 417, in load_modules
    force, status, report, loaded_modules, update_module, models_to_check)
  File "/opt/odoo12/odoo/modules/loading.py", line 313, in load_marked_modules
    perform_checks=perform_checks, models_to_check=models_to_check
  File "/opt/odoo12/odoo/modules/loading.py", line 222, in load_module_graph
    load_data(cr, idref, mode, kind='data', package=package, report=report)
  File "/opt/odoo12/odoo/modules/loading.py", line 68, in load_data
    tools.convert_file(cr, package.name, filename, idref, mode, noupdate, kind, report)
  File "/opt/odoo12/odoo/tools/convert.py", line 801, in convert_file
    convert_xml_import(cr, module, fp, idref, mode, noupdate, report)
  File "/opt/odoo12/odoo/tools/convert.py", line 864, in convert_xml_import
    obj.parse(doc.getroot(), mode=mode)
  File "/opt/odoo12/odoo/tools/convert.py", line 763, in parse
    exc_info[2]
  File "/opt/odoo12/odoo/tools/pycompat.py", line 86, in reraise
    raise value.with_traceback(tb)
  File "/opt/odoo12/odoo/tools/convert.py", line 757, in parse
    self._tags[rec.tag](rec, de, mode=mode)
  File "/opt/odoo12/odoo/tools/convert.py", line 662, in _tag_record
    record = model.with_context(rec_context)._load_records([data], self.mode == 'update')
  File "/opt/odoo12/odoo/models.py", line 3788, in _load_records
    data['record'].write(data['values'])
  File "/opt/odoo12/addons/website/models/ir_ui_view.py", line 41, in write
    return super(View, self).write(vals)
  File "/opt/odoo12/odoo/addons/base/models/ir_ui_view.py", line 441, in write
    return super(View, self).write(self._compute_defaults(vals))
  File "/opt/odoo12/odoo/models.py", line 3306, in write
    fields[0].determine_inverse(records)
  File "/opt/odoo12/odoo/fields.py", line 1097, in determine_inverse
    getattr(records, self.inverse)()
  File "/opt/odoo12/odoo/addons/base/models/ir_ui_view.py", line 275, in _inverse_arch
    view.write(data)
  File "/opt/odoo12/addons/website/models/ir_ui_view.py", line 41, in write
    return super(View, self).write(vals)
  File "/opt/odoo12/odoo/addons/base/models/ir_ui_view.py", line 441, in write
    return super(View, self).write(self._compute_defaults(vals))
  File "/opt/odoo12/odoo/models.py", line 3269, in write
    self._write(store_vals)
  File "/opt/odoo12/odoo/models.py", line 3413, in _write
    self._validate_fields(vals)
  File "/opt/odoo12/odoo/models.py", line 1112, in _validate_fields
    raise ValidationError("%s\n\n%s" % (_("Error while validating constraint"), tools.ustr(e)))
odoo.tools.convert.ParseError: "Error mientras se validaban las restricciones

El campo `project_department_id` no existe

Contexto del error:
Vista `project.task.department.form`
[view_id: 2138, xml_id: project_department.view_task_form2, model: project.task, parent_id: 403]
None" while parsing /opt/odoo12/addons_extra/project_department/views/project.xml:58, near
<record id="view_task_form2" model="ir.ui.view">
        <field name="name">project.task.department.form</field>
        <field name="model">project.task</field>
        <field name="inherit_id" ref="project.view_task_form2"/>
        <field eval="2" name="priority"/>
        <field name="arch" type="xml">
            <field name="partner_id" position="after">
              <field name="project_department_id"/>
            </field>
        </field>
    </record>
```
When I tried to install for the first time I got "KeyError: 'department_id'", then I just commented all related to department_id, just to install the module. Once installed, I uncommented all references back. Finally when trying to upgrade the module I get the error above.

**Comment by Nikul-Chaudhary** [comment]
> Opps, It's Base on [OCA/account-analytic#209](https://github.com/OCA/account-analytic/pull/209)

> So fields not get, But you can review only Code, If base modules are Done then testing in runbot


@xabispacebiker I already Inform that. Thank you

**Comment by ovnicraft** [comment]
@Nikul-Chaudhary is not clear your comment, can you update here please.

**Comment by emagdalenaC2i** [comment]
@ovnicraft I think @Nikul-Chaudhary comment means that this module depends on OCA/account-analytic#209 that isn't yet merged.

@pedrobaeza OCA/account-analytic#20 is already approved, could you review it?

**Comment by pedrobaeza** [comment]
Please rebase

**Comment by ovnicraft** [comment]
@Nikul-Chaudhary  `department_id` field is not defined on `project.project`and views got it. Please add it.

**Comment by Nikul-Chaudhary** [comment]
> @Nikul-Chaudhary `department_id` field is not defined on `project.project`and views got it. Please add it.

@ovnicraft In v11 `Project class` inherited  `account.analytic.account ` like this `_inherits = {'account.analytic.account': "analytic_account_id"}` [check here](https://github.com/odoo/odoo/blob/11.0/addons/project/models/project.py#L50)
and we added field here https://github.com/OCA/account-analytic/pull/209/commits/46cfac0a148cd5464fdcc4a31c44ce34756ac26f#diff-4412e57269867d97a5c8559495ce4f25R12

Now In v12 `_inherits` is removed in `Project class` , That's why fatching Error like `KeyError: 'department_id'`


@pedrobaeza @dreispt @emagdalenaC2i Cann you please help here what should I Do?

I adding code like this or any Other way?
 ```
class Project(models.Model):
    _name = "project.project"
    _description = "Project"
    _inherit = ['mail.alias.mixin', 'mail.thread', 'portal.mixin']
    _inherits = {'account.analytic.account': "analytic_account_id"}
```

**Comment by ovnicraft** [comment]
@Nikul-Chaudhary so you can depends `analytic_base_department` and add inherits in project like you mention, but i recommend you a new module to reinsert `_inherits = {'account.analytic.account': "analytic_account_id"}`

So there is another PR adding again `_inherits` https://github.com/OCA/project/pull/471 here: [diff](https://github.com/OCA/project/pull/471/files?file-filters%5B%5D=.py&file-filters%5B%5D=.xml#diff-836e3a974ffa8d76a81bfa20c5111dfaR16)

**Comment by bjeficent** [comment]
Hi @ovnicraft, as you said in that PR I added again to avoid conflicts and may be adding a new module to reinsert ```_inherits``` and make these dependent of it should be better.

**Comment by ovnicraft** [comment]
@bjeficent yes could be `project_analytic` if nobody does it until weekend i can work on it.
Stay tuned for updates.

**Comment by Nikul-Chaudhary** [comment]
@pedrobaeza @ovnicraft @bjeficent @aheficent How can I solve this problem: https://github.com/OCA/project/pull/465#issuecomment-487402758

Can you please help here??

**Comment by pedrobaeza** [comment]
I would say that you only need to add a related field pointing to `account_analytic_id.department_id` (or the names they have).

**Comment by rafaelbn** [comment]
Hi @Nikul-Chaudhary , thanks for the migration. This PR is waiting for you fix please :smile:

**Comment by Nikul-Chaudhary** [comment]
> Hi @Nikul-Chaudhary , thanks for the migration. This PR is waiting for you fix please

When I have time, I am working on it.

Thank for remainder

**Pull request closed without merging**